persistence-cassandra-4.0/src/main/java/io/stargate/db/cassandra/impl/StargateQueryHandler.java

```diff
@@ -244,69 +258,228 @@ private void authorizeByToken(ByteBuffer token, CQLStatement statement) {
+            String.format("Missing correct permission on role %s", resource), e);
       }
 
       logger.debug(
-          "preparing to authorize statement of type {} on {}.{}",
+          "authorized statement of type {} on {}", castStatement.getClass().toString(), resource);
+    } else if (statement instanceof ListRolesStatement) {
+      ListRolesStatement stmt = (ListRolesStatement) castStatement;
+      String role = getRoleResourceFromStatement(stmt, "grantee");
+      logger.debug(
+          "preparing to authorize statement of type {} on {}",
           castStatement.getClass().toString(),
-          keyspaceName,
-          tableName);
+          role);
 
       try {
-        authorization.authorizeSchemaWrite(authToken, keyspaceName, tableName, scope);
+        authorization.authorizeRoleRead(authToken, role);
       } catch (io.stargate.auth.UnauthorizedException e) {
         throw new UnauthorizedException(
-            String.format(
-                "Missing correct permission on %s.%s",
-                keyspaceName, (tableName == null ? "" : tableName)));
+            String.format("Missing correct permission on role %s", role), e);
       }
 
       logger.debug(
-          "authorized statement of type {} on {}.{}",
+          "authorized statement of type {} on {}", castStatement.getClass().toString(), role);
+    } else if (statement instanceof ListPermissionsStatement) {
+      ListPermissionsStatement stmt = (ListPermissionsStatement) castStatement;
+      String role = getRoleResourceFromStatement(stmt, "grantee");
+      logger.debug(
+          "preparing to authorize statement of type {} on {}",
           castStatement.getClass().toString(),
-          keyspaceName,
-          tableName);
+          role);
+
+      try {
+        authorization.authorizePermissionRead(authToken, role);
+      } catch (io.stargate.auth.UnauthorizedException e) {
+        throw new UnauthorizedException(
+            String.format("Missing correct permission on role %s", role), e);
+      }
+
+      logger.debug(
+          "authorized statement of type {} on {}", castStatement.getClass().toString(), role);
+    }
+  }
+
+  private String getRoleResourceFromStatement(Object stmt, String fieldName) {
+    try {
+      Class<?> aClass = stmt.getClass();
+      if (stmt instanceof ListUsersStatement
+          || stmt instanceof GrantPermissionsStatement
+          || stmt instanceof RevokePermissionsStatement) {
+        aClass = aClass.getSuperclass();
+      }
+
+      Field f = aClass.getDeclaredField(fieldName);
+      f.setAccessible(true);
+      RoleResource roleResource = (RoleResource) f.get(stmt);
+
+      return roleResource != null ? roleResource.getName() : null;
+    } catch (Exception e) {
+      logger.error("Unable to get " + fieldName, e);
+      throw new RuntimeException("Unable to get private field", e);
     }
   }
 
+  private String getResourceFromStatement(PermissionsManagementStatement stmt) {
```

Comment:
It'd be nice if these reflection base methods were be added to `ReflectionUtils.java`. I'm wondering if we should statically initialize the `Field` too so it doesn't have to be done for every request.

Reply:
I'm not sure I follow about statically initializing `Field`

Reply:
Never mind on this. We're getting the fields of types that aren't know at compile time.